### DISCUSSION TITLE
trait `fltk::prelude::WidgetExt` is not implemented for custom widget with fl2rust

### DISCUSSION THREAD

**Discussion by martinellison** [fltk-rs/fltk-rs#1606] [Q&A] [answered]
* I have implemented a custom widget 
* I have added the new custom widget to a group in FLUID
* I have generated code using fl2rust from my FLUID file
* and I get this error that WidgetExt "is not implemented"

The fl2rust-generated code has:

```
	let mut my_widget = MyWidget::new(15, 30, 410, 320, None);
        ...
	tree_group.add(&node_tree);
        tree_group.resizable(&node_tree);
```
and the error is against the calls to `add` and `resizable`.

I had to implement `MyWidget::new` to get the call to build.

My code has:

```
widget_extends!(MyWidget, Widget, inner);
pub struct MyWidget {
    inner: Widget,
   // more fields
}
```
which, as far as I can see, follows the documentation (https://fltk-rs.github.io/fltk-book/Custom-Widgets.html).

Could anyone please explain what I am doing wrong and what I should be doing?

**Comment by MoAlyousef** [comment, marked as answer]
Hi
Since Rust doesn't have inheritance, this is emulated using the deref pattern. i.e. you would need to:
```rust
        tree_group.add(&*node_tree);
        tree_group.resizable(&*node_tree);
```
AFAIK the fluid gui doesn't allow you to add your custom widget via the gui, so it would have to be done in code.

**Comment by martinellison** [comment]
Ok, thanks for clarifying.

(edit) also, it seems that I need to declare `Deref` to make it work: `use std::ops::Deref;`
